File: tinkoff/contest 3 (Деревья)/B.cpp
// B.cpp : Этот файл содержит функцию "main". Здесь начинается и заканчивается выполнение программы.
//

#include <iostream>

using namespace std;

struct ans {
	int deep;
	int maxi;
	int mini;
};

ans check_tree(int tree[][2], int n) {
	if ((tree[n][0] == -1) and (tree[n][1] == -1)) {
		ans ret; ret.deep = 0; ret.maxi = n; ret.mini = n;
		return { 0, n, n };
	}
	int deep = 0;
	int maxi = n;
	int miniLeft = n;
	if (tree[n][0] != -1) {
		int left = tree[n][0];
		ans st = check_tree(tree, left);
		deep = st.deep; maxi = st.maxi;
		miniLeft = st.mini;
		if (maxi >= n) {
			return { -1, -1, -1 };
		}
	}
	maxi = max(n, maxi);
	int deep2 = 0;
	int mini = n;
	int maxiRight = n;
	if (tree[n][1] != -1) {
		int right = tree[n][1];
		ans st = check_tree(tree, right);
		deep2 = st.deep; mini = st.mini;
		maxiRight = st.maxi;
		if (mini <= n) {
			return { -1, -1, -1 };
		}
	}
	mini = min(n, mini);
	if ((deep != -1) and (deep2 != -1) and (abs(deep - deep2) <= 1)) {
		return { max(deep, deep2) + 1, maxiRight, mini };
	}
	return { -1, -1, -1 };
}


int main()
{
	int n, k;
	cin >> n >> k;
	int tree[100001][2];
	for (int i = 0; i < n; i++) {
		cin >> tree[i][0] >> tree[i][1];
	}

	ans ans1 = check_tree(tree, k);
	if (ans1.deep == -1) {
		cout << 0;
	}
	else {
		cout << 1;
	}
}

File: tinkoff/contest 3 (Деревья)/B.py
import sys
sys.setrecursionlimit(100001)
def check_tree(tree, n):
    if (tree[n] == (-1, -1)):
        return (True, 0, n, n)
    ans, deep, maxi = True, 0, n
    if (tree[n][0] != -1):
        left = tree[n][0]
        ans, deep, maxi, mini = check_tree(tree, left)
        if maxi >= n:
            return (False, -1, -1, -1)
    maxi = max(n, maxi)
    
    ans2, deep2, mini = True, 0, n
    if (tree[n][1] != -1):
        right = tree[n][1]
        ans2, deep2, maxi, mini = check_tree(tree, right)
        if mini <= n:
            return (False, -1, -1, -1)
    mini = min(n, mini)
    if (ans and ans2 and abs(deep - deep2) <= 1):
        return (True, max(deep, deep2)+1, maxi, mini)
    return (False, -1, -1, -1)


n, k = map(int, sys.stdin.readline().split())
tree = {i:(-1, -1) for i in range(n)}

for i in range(n):
    a, b = map(int, sys.stdin.readline().split())
    tree[i] = (a, b)

sys.stdout.write(str(int(check_tree(tree, k)[0])))
